mlx_graphs/nn/conv/rel_conv.py

```diff
@@ -0,0 +1,327 @@
+
+
+class GeneralizedRelationalConv(MessagePassing):
+    r"""Generalized relational convolution layer from
+    `"Neural Bellman-Ford Networks: A General Graph Neural Network Framework
+      for Link Prediction" <https://arxiv.org/abs/2106.06935>`_ paper.
+
+    Adopted from the PyG version from here:
+    `https://github.com/KiddoZhu/NBFNet-PyG/blob/master/nbfnet/layers.py`
+
+    Part of the Neural Bellman-Ford networks (NBFNet) holding
+    state-of-the-art in KG completion.
+    Works with multi-relational graphs where edge types are stored in `edge_labels`.
+    The message function composes node and relation vectors in three possible ways.
+    The expected behavior is to work with "labeling trick" graphs where one node
+    in the graph is labeled with a `query` vector, while rest are zeros.
+    Message passing is then done separately for each data point in the batch.
+    The input shape is expected to be [batch_size, num_nodes, input_dim]
+
+    Alternatively, the layer can work as a standard relational conv
+    with shapes [num_nodes, input_dim].
+
+    Note that this implementation materializes all edge messages and is O(E).
+    The complexity can be further reduced by adopting the O(V) `rspmm` C++ kernel
+    from the NBFNet-PyG repo to the MLX framework (not implemented here).
+
+    Args:
+        input_dim: input feature dimension (same for node and edge features)
+        output_dim: output node feature dimension
+        num_relation: number of unique relations in the graph
+        message_func: "transe" (sum), "distmult" (mult),
+            "rotate" (complex rotation). Default: ``distmult``
+        aggregate_func: "add", "mean", or "pna". Default: ``add``
+        layer_norm: whether to use layer norm
+            (often crucial to the performance). Default: ``True``
+        activation: non-linearity. Default: ``relu``
+        dependent: whether to use separate relation embedding matrix
+             ``False`` or build relations from the input relations ``True``
+        node_dim: for 3D batches, specified which dimension contains all nodes.
+            Default: ``0``
+
+    Example:
+
+    .. code-block:: python
+
+        import mlx.core as mx
+        import mlx.nn as nn
+        from mlx_graphs.nn import GeneralizedRelationalConv
+
+        input_dim = 16
+        output_dim = 16
+        num_relations = 3
+
+        conv = GeneralizedRelationalConv(input_dim, output_dim, num_relations)
+
+        batch_size = 2
+        edge_index = mx.array([[0, 1, 2, 3, 4], [0, 0, 1, 1, 3]])
+        edge_types = mx.array([0, 0, 1, 1, 2])
+        boundary = mx.random.uniform(0, 1, shape=(batch_size, 5, 16)
+        size = (node_features.shape[0], node_features.shape[0])
+
+        layer_input = boundary
+        h = conv(layer_input, query, boundary, edge_index, edge_type, size)
+
+        # optional: residual connection if input dim == output dim
+        h = h + layer_input
+        layer_input = h
```

Comment:
The given example has some errors. I just pushed a fix for the layer import. It remains to add the `node_features` and `query` arrays, which are missing.

Reply:
fixed the docstring a little https://github.com/mlx-graphs/mlx-graphs/pull/111/commits/63e55b84d9d736e5ff717955c214c0c7c672130e